Roadmap.md

```diff
@@ -243,3 +243,151 @@ Fetch API:
 ```ts
 fetch(input?: Request | string, init?: RequestInit): Promise<Response>;
 ```
+
+#### I/O
+
+There are many OS constructs that perform I/O: files, sockets, pipes.
+Deno aims to provide a unified lowest common denominator interface to work with
+these objects. Deno needs to operate on all of these asynchronously in order
+to not block the event loop and it.
+
+Sockets and pipes support non-blocking reads and write.  Generally file I/O is
+blocking but it can be done in a thread pool to avoid blocking the main thread.
+Although file I/O can be made asynchronous, it does not support the same
+non-blocking reads and writes that sockets and pipes do.
+
+The following interfaces support files, socket, and pipes and are heavily
+inspired by Go. The main difference in porting to JavaScript is that errors will
+be handled by exceptions, modulo EOF, which is returned as part of
+`ReadResult`.
+
+```ts
+// The bytes read during an I/O call and a boolean indicating EOF.
+interface ReadResult {
+  nread: number;
+  eof: boolean;
+}
+
+// Reader is the interface that wraps the basic Read method.
+// https://golang.org/pkg/io/#Reader
+interface Reader {
+  // Read reads up to p.byteLength bytes into p. It returns the number of bytes
+  // read (0 <= n <= p.byteLength) and any error encountered. Even if Read
+  // returns n < p.byteLength, it may use all of p as scratch space during the
+  // call. If some data is available but not p.byteLength bytes, Read
+  // conventionally returns what is available instead of waiting for more.
+  //
+  // When Read encounters an error or end-of-file condition after successfully
+  // reading n > 0 bytes, it returns the number of bytes read. It may return the
+  // (non-nil) error from the same call or return the error (and n == 0) from a
+  // subsequent call. An instance of this general case is that a Reader
+  // returning a non-zero number of bytes at the end of the input stream may
+  // return either err == EOF or err == nil. The next Read should return 0, EOF.
+  //
+  // Callers should always process the n > 0 bytes returned before considering
+  // the error err. Doing so correctly handles I/O errors that happen after
+  // reading some bytes and also both of the allowed EOF behaviors.
+  //
+  // Implementations of Read are discouraged from returning a zero byte count
+  // with a nil error, except when p.byteLength == 0. Callers should treat a
+  // return of 0 and nil as indicating that nothing happened; in particular it
+  // does not indicate EOF.
+  //
+  // Implementations must not retain p.
+  async read(p: Uint8Array): Promise<ReadResult>;
```

Comment:
1.  Maybe we should say `p: ArrayBufferView`, where 
    ```ts
    interface ArrayBufferView {
      buffer: ArrayBuffer,
      byteOffset: number,
      byteLength: number
    }
    ```

2.  I feel that `p` should be optional. Reading into a preallocated buffer is probably optimal if you can re-use the same buffer over and over again. But I think a lot of code will end up doing this:
    ```ts
    let buf = new Uint8Array(64436);
    const r = conn.read(buf);
    r = r.subarray(0, nread);
    ```

Reply:
+1 on ArrayBufferView

-1 ln optional arguments. This is meant to be the most minimal interface that objects must implement to be a Reader - this ergonomics are left to higher layers.

Reply:
+1 on ArrayBufferView too.

If we expose `Symbol.asyncIterator` around `.read` that works with a view (and hopefully no buffer allocations at all but just a view into the memory) we get the minimal `.read` as well as the usable `for await`.